rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -796,7 +797,99 @@ void main_window::InstallPup(QString file_path)
 	}
 }
 
-void main_window::HandlePupInstallation(QString file_path)
+void main_window::ExtractPup()
+{
+	const QString path_last_pup = m_gui_settings->GetValue(gui::fd_extract_pup).toString();
+	QString file_path = QFileDialog::getOpenFileName(this, tr("Select PS3UPDAT.PUP To extract"), path_last_pup, tr("PS3 update file (PS3UPDAT.PUP);;All pup files (*.pup);;All files (*.*)"));
+
+	QString dir = file_path.isEmpty() ? QString{} : QFileDialog::getExistingDirectory(this, tr("Open Directory"), QString{}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
+
+	if (!dir.isEmpty())
+	{
+		HandlePupInstallation(file_path, dir);
+	}
+}
+
+void main_window::ExtractTar()
+{
+	const QString path_last_pup = m_gui_settings->GetValue(gui::fd_extract_tar).toString();
+	QString file_path = QFileDialog::getOpenFileName(this, tr("Select TAR To extract"), path_last_pup, tr("All tar files (*.tar *.tar.aa.*);;All files (*.*)"));
+
+	QString dir = file_path.isEmpty() ? QString{} : QFileDialog::getExistingDirectory(this, tr("Open Directory"), QString{}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
+
+	if (!dir.isEmpty())
+	{
+		m_gui_settings->SetValue(gui::fd_extract_tar, QFileInfo(file_path).path());
+		HandleTarExtraction(file_path, dir);
+	}
+}
+
+void main_window::HandleTarExtraction(QString file_path, QString dir_path)
```

Comment:
same as mself.
There's no reason to implement an extraction in a GUI window.
This is:
A. Bloating the GUI code.
B. Not portable if someone ever decided to change the frontend.
C. Not necessary as there is no progress dialog attached. (It should ideally support some sort of events if ever necessary)

Reply:
Okay, I made it not in main_window class but a static function. The reason for it is that main_window.cpp already includes both unself.h and TAR.h for other code.